src/storage-client/src/controller.rs

```diff
@@ -1595,9 +1595,28 @@ where
             Some(StorageResponse::FrontierUppers(updates)) => {
                 self.update_write_frontiers(&updates);
             }
-            Some(StorageResponse::DroppedIds(_ids)) => {
-                // TODO(petrosagg): It looks like the storage controller never cleans up GlobalIds
-                // from its state. It should probably be done as a reaction to this response.
+            Some(StorageResponse::DroppedIds(ids)) => {
```

Comment:
I have a suspicion that these IDs might be a mix of "top-level" IDs and sub-source IDs (for PG sources). Will the below code work for that? What's the remap shard of a sub source (I'm wondering because they all should share the remap shard of the top-level source)

Reply:
I believe this is acceptable for both sources and subsources. The remap shard of a subsource is empty as all non-ingestion-based sources are. The PG source's remap shard gets used, and that source simply demuxes the data it ingests into the data shards of the subsources.

If I understand the apprehension, the shards that are getting retired here are each unique so there's no fear of finalizing someone else's remap shard from underneath them. This must be the case because the source's subsources get created before the source itself and we undergo no effort to rewrite the remap shards of the subsources to be the same as the source.

If this doesn't 100% check out to you can spend some time in the subsource code to see if I can find the spot where the demuxing happens without consulting the subsources' remap shards.

Reply:
I looked into this: we only get `DroppedIds` messages for the top-level source. You can double-check by adding a log statement here and creating/dropping a `LOAD GENERATOR TPCH` source.

It seems we need to look into the `CollectionState` (via `collection()`) to see what the source exports are.

There's another problem here, which @guswynn is already aware of. We often don't get a `DroppedIds` at all, mostly when using linked clusters because the cluster/connection is torn down before we get a chance to report back the `DroppedIds` message.

Reply:
We still don't drop the shards for all exports?

Reply:
This was done in #17687, which just already merged.

Reply:
my bad! 🙈